### ISSUE TITLE
package:glob is very slow

### ISSUE THREAD

**Issue by scheglov** [dart-lang/sdk#24962]
@nex3 @munificent @bwilkerson 

The benchmark below takes `240 ms` using `package:glob` and `2.17 ms` when uses [Analyzer's implementation.](https://github.com/dart-lang/sdk/blob/master/pkg/analyzer/lib/src/util/glob.dart) This is `110` times slower.

```
import 'package:benchmark_harness/benchmark_harness.dart';

import 'glob.dart' as gm;
import 'package:glob/glob.dart' as gp;
import 'dart:async';

main() {
  GlobBenchmark.main();
  new Future.delayed(new Duration(days: 1));
}

class GlobBenchmark extends BenchmarkBase {
//  gm.Glob glob = new gm.Glob('/','**/*.dart');
  var glob = new gp.Glob('/**/*.dart');

  GlobBenchmark() : super('GlobBenchmark');

  void run() {
    bool b1 = glob.matches('/Users/scheglov/dart/mdl/lib/mdl.dart');
    bool b3 = glob.matches('/Users/scheglov/dart/mdl/lib/mdl.html');
    assert(b1);
    assert(!b3);
  }

  void setup() {}

  void teardown() {}

  static void main() {
    new GlobBenchmark().report();
  }
}
```

![image](https://cloud.githubusercontent.com/assets/384794/11226198/41920a84-8d33-11e5-92de-892ba7562dcc.png)

![image](https://cloud.githubusercontent.com/assets/384794/11226209/55a41742-8d33-11e5-82ae-8ab33024136f.png)

**Comment by nex3** [comment]
Your benchmark file imports an `absolute_path.dart` file that I don't know about. Can you provide a benchmark that's runnable in the context of `glob`?

**Comment by scheglov** [comment]
I've cleaned the benchmark out.

**Comment by nex3** [comment]
[This CL](https://codereview.chromium.org/1455003002/) optimizes `path.absolute()` for the case where paths are already absolute, and `path.normalize()` for the case where the paths don't need to be normalized. On my machine, it trims the glob package's speed down to about 9ms relative to the analyzer implementation's 6ms.

Is this acceptable performance for the analysis server? The only way I can see to get it faster would be to remove the normalization logic entirely, but that would be a breaking change that would negatively affect the package's general usability, so I'd much rather not do it.

**Comment by nex3** [comment]
Alternately, you could create [a custom `path` package context that never normalizes paths](https://gist.github.com/nex3/ae2ecc16834cea24e265). This is somewhat off the beaten path, but on my benchmarks it (combined with the `path` optimizations) makes the `glob` package implementation about 1us faster than the analyzer implementation.

**Comment by scheglov** [comment]
Nice!

With this CL `package:glob` is just `3` times slower that the analyzer's implementation in the benchmark - `7.65 us` vs. `2.09 us`. Would special-casing `**/*.dart` (i.e. match any Dart file anywhere, on Linux or Windows) as a suffix matching make it on-par with the analyzer's implementation?

**Comment by nex3** [comment]
> With this CL `package:glob` is just `3` times slower that the analyzer's implementation in the benchmark - `7.65 us` vs. `2.09 us`.

That's very different than the numbers I'm seeing (9us vs 6us). Are you still using the benchmark above? What Dart version are you using?

What happens if you run with the custom `path` context I posted above?

> Would special-casing `**/*.dart` (i.e. match any Dart file anywhere, on Linux or Windows) as a suffix matching make it on-par with the analyzer's implementation?

I'm not sure what you mean by special-casing. Keep in mind that most of the overhead (at least as far as I can measure) comes from the check for whether we should normalize a path; unless we can dodge that check somehow, I don't think we can get much faster.

**Comment by scheglov** [comment]
Yes, I use the same benchmark.
I use this version of analyzer's Glob.
https://github.com/dart-lang/sdk/commit/96b494665e0aeb6e3b7ed37bf3db52d4c50cbfc9

With `NoNormalizeContext` it is a little faster `5.62 us`, but still slower then the analyzer's version.

The `**/*.dart` pattern can be implemented as just `path.endsWith('.dart')` without any AST and without any call to other methods.

**Comment by nex3** [comment]
The `/` in `**/*.dart` means that it only matches paths that are at least one directory deep. This means that the path `"foo.dart"` doesn't match, even though it passes `path.endsWith('.dart')`. If you want *any* relative path ending in `.dart`, you should use `**.dart`.

**Comment by scheglov** [comment]
Actually in DAS we mostly apply globs to absolute paths.
If `**.dart` works for absolute paths and we can optimize it - great.

If it does not work, then it should :-)
See https://github.com/dart-lang/glob/issues/2

This way it would be both work faster and easier to implement.

**Comment by nex3** [comment]
> Actually in DAS we mostly apply globs to absolute paths.

Is this a flexible constraint? It seems to be causing a lot of pain. As I understand it, the globs you're dealing with are only ever relevant within packages, so why not just run them against the paths relative to those packages? Forcing them to be absolute brings in all the complexity of drive letters, and makes intentionally-relative globs like `test/**.dart` not work how people expect (without a relative context, that would just as happily match match `/.../pkg/lib/src/test/...` as `/.../pkg/test/...`).

**Comment by scheglov** [comment]
We list files (using our VFS) and get their absolute paths.
Then we check whether we know how to analyze them.
By default we know how to analyze and Dart or HTML files.
And we don't actually care about path, but we need to make it general enough for plugins to DAS.
So, we use globs and by default have `**.dart` and `**.html`.

We *also* apply globs to relative path, for example to exclude analysis of examples `examples/**`.
But in this case we know that target paths are relative.

Basically - don't make decisions for me because you have also listing in your `Glob` implementation. We don't need this in DAS. Just make it simple and make `**` to match just anything.

**Comment by nex3** [comment]
It doesn't matter whether the paths you're passing in are relative or absolute—`glob` looks at their notional location, not their textual contents. What matters is the working directory as encoded in the context you pass to the glob. So using your example paths from #7, you could use the glob `"**.dart"` relative to `/Users/scheglov/dart/dart-material-design-lite` and avoid all sorts of absolute-path complexity.

> Basically - don't make decisions for me because you have also listing in your `Glob` implementation. We don't need this in DAS. Just make it simple and make `**` to match just anything.

Part of existing in an ecosystem is making sacrifices to one use case in order to support others. Another part is making decisions that make sense at a large scale, even if they make smaller-scale stuff more difficult. Being able to use the same glob syntax and semantics across the Dart ecosystem is very valuable for our users, and part of what that means is that `**` should mean the same thing everywhere—in DAS as well as in places where listing is used.

**Comment by bwilkerson** [comment]
> ... part of what that means is that ** should mean the same thing everywhere ...

Unfortunately, it's too late for that, given that the quiver package already implements a different semantic than the glob package.

**Comment by nex3** [comment]
As I understand it, quiver is currently unmaintained and we don't encourage people to use it.

**Comment by sethladd** [comment]
cc @yjbanov @cbracken re: quiver

**Comment by cbracken** [comment]
Quiver is actively maintained by both Yegor and myself. We're currently in the middle of a refactor to split it up into a pile of smaller packages, which are hosted under the [QuiverDart](https://github.com/QuiverDart/) org. There are >100 commits in the last 3 weeks.

**Comment by yjbanov** [comment]
We considered deprecating Quiver's glob implementation, but had to keep it due to performance concerns about the `package:glob` implementation.

See: https://github.com/google/quiver-dart/issues/207

**Comment by munificent** [comment]
Do you have any benchmarks?

**Comment by yjbanov** [comment]
@munificent: no (assuming you are asking about quiver benchmarks)

**Comment by munificent** [comment]
Closing this old bug. If there are still perf concerns, please file a new bug here: https://github.com/dart-lang/glob.
